README.md
# f1-telemetry-dashboard

This template should help get you started developing with Vue 3 in Vite.

## Recommended IDE Setup

[VSCode](https://code.visualstudio.com/) + [Volar](https://marketplace.visualstudio.com/items?itemName=Vue.volar) (and disable Vetur) + [TypeScript Vue Plugin (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.vscode-typescript-vue-plugin).

## Type Support for `.vue` Imports in TS

TypeScript cannot handle type information for `.vue` imports by default, so we replace the `tsc` CLI with `vue-tsc` for type checking. In editors, we need [TypeScript Vue Plugin (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.vscode-typescript-vue-plugin) to make the TypeScript language service aware of `.vue` types.

If the standalone TypeScript plugin doesn't feel fast enough to you, Volar has also implemented a [Take Over Mode](https://github.com/johnsoncodehk/volar/discussions/471#discussioncomment-1361669) that is more performant. You can enable it by the following steps:

1. Disable the built-in TypeScript Extension
    1) Run `Extensions: Show Built-in Extensions` from VSCode's command palette
    2) Find `TypeScript and JavaScript Language Features`, right click and select `Disable (Workspace)`
2. Reload the VSCode window by running `Developer: Reload Window` from the command palette.

## Customize configuration

See [Vite Configuration Reference](https://vitejs.dev/config/).

## Project Setup

```sh
npm install
```

### Compile and Hot-Reload for Development

```sh
npm run dev
```

### Type-Check, Compile and Minify for Production

```sh
npm run build
```

### Run Unit Tests with [Vitest](https://vitest.dev/)

```sh
npm run test:unit
```

### Lint with [ESLint](https://eslint.org/)

```sh
npm run lint
```


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/Pehesi97/f1-telemetry-dashboard/compare/v1.0.4...v1.1.0) (2023-12-11)


### Features

* more session data ([d2c1977](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/d2c197709462ea6a2e2d6ca6b89b0b72a3eee726))
* random stuff ([fce28b9](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/fce28b9df0cbeff60f08977257746a79a4854dee))


### Bug Fixes

* fraction digits ([9a09161](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/9a09161572fbe8f47654c7860ab525cee8a65be9))

## [1.0.4](https://github.com/Pehesi97/f1-telemetry-dashboard/compare/v1.0.3...v1.0.4) (2023-09-14)


### Bug Fixes

* testing get PRs ([#10](https://github.com/Pehesi97/f1-telemetry-dashboard/issues/10)) ([8e2b11b](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/8e2b11bad9952123ece9bdab8f69879578cbed0f))

## [1.0.3](https://github.com/Pehesi97/f1-telemetry-dashboard/compare/v1.0.2...v1.0.3) (2023-09-14)


### Bug Fixes

* workflow ([#8](https://github.com/Pehesi97/f1-telemetry-dashboard/issues/8)) ([be7fca4](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/be7fca40da95c776201980576c30bcbcaf3bbf95))

## [1.0.2](https://github.com/Pehesi97/f1-telemetry-dashboard/compare/v1.0.1...v1.0.2) (2023-09-14)


### Bug Fixes

* test release please ([#6](https://github.com/Pehesi97/f1-telemetry-dashboard/issues/6)) ([004cae2](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/004cae22e118c5fde7979649aeaf4ff25ffd8fa6))

## [1.0.1](https://github.com/Pehesi97/f1-telemetry-dashboard/compare/v1.0.0...v1.0.1) (2023-09-14)


### Bug Fixes

* debug release-please ([#4](https://github.com/Pehesi97/f1-telemetry-dashboard/issues/4)) ([4c6c7eb](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/4c6c7eb87f9fb19ee416e5d07b9d229f5299c0ca))

## 1.0.0 (2023-09-14)


### Features

* menu ([#1](https://github.com/Pehesi97/f1-telemetry-dashboard/issues/1)) ([b823493](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/b823493d4da0a5f325871faac092c890533d1a88))
* using TimeChart ([f775a54](https://github.com/Pehesi97/f1-telemetry-dashboard/commit/f775a54ff5f80db27dedbd354db3fea6d56c6567))


backend/app.ts
import { F1TelemetryClient, constants } from '@racehub-io/f1-telemetry-client';
import { PacketLapData, PacketCarTelemetryData, PacketSessionData, PacketSessionHistoryData, PacketLobbyInfoData, PacketCarStatusData, PacketParticipantsData } from '@racehub-io/f1-telemetry-client/build/main/parsers/packets/types';

const { PACKETS } = constants;

import http from 'http';
import { Server, Socket } from "socket.io";

const server: http.Server = http.createServer();
server.listen(3000, "0.0.0.0");

const io: Server = new Server(server, {
  cors: {
    origin: "http://localhost:5173"
  }
});

io.on('connection', (socket: Socket) => {
  console.log('a user connected');
});

const client: F1TelemetryClient = new F1TelemetryClient({ port: 20777, bigintEnabled: false });
client.start();



client.on(PACKETS.lapData, (data: PacketLapData) => {
  io.emit("lapData", {
    position: data.m_lapData[data.m_header.m_playerCarIndex].m_carPosition,
    lastLapTime: data.m_lapData[data.m_header.m_playerCarIndex].m_lastLapTimeInMS,
    currentLapTime: data.m_lapData[data.m_header.m_playerCarIndex].m_currentLapTimeInMS,
    sector1Time: data.m_lapData[data.m_header.m_playerCarIndex].m_sector1TimeMinutes,
    sector2Time: data.m_lapData[data.m_header.m_playerCarIndex].m_sector2TimeMinutes,
    currentLapNum: data.m_lapData[data.m_header.m_playerCarIndex].m_currentLapNum
  });
});

client.on(PACKETS.carTelemetry, (data: PacketCarTelemetryData) => {
  io.emit("telemetryData", {
    throttle: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_throttle,
    brake: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_brake,
    gear: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_gear,
    speed: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_speed,
    frame: data.m_header.m_frameIdentifier,
  });
});

client.on(PACKETS.carStatus, (data: PacketCarStatusData) => {
  let tyreWearData = [] as any[]
  data.m_carStatusData.forEach((car) => {
    tyreWearData.push({
      a: car.m_tyresWear?.[0] ?? 0,
      b: car.m_tyresWear?.[1] ?? 0,
      c: car.m_tyresWear?.[2] ?? 0,
      d: car.m_tyresWear?.[3] ?? 0,
    })
  });
  io.emit('tyreWearData', {
    data: tyreWearData
  })
})

client.on(PACKETS.participants, (data: PacketParticipantsData) => {
  io.emit('drivers', {
    drivers: data.m_participants,
    driverCount: data.m_numActiveCars
  })
})

client.on(PACKETS.session, (data: PacketSessionData) => {
  io.emit("sessionData", {
    aiDifficulty: data.m_aiDifficulty,
    track: constants.TRACKS[data.m_trackId].name,
    weather: data.m_weather,
  });
});

client.on(PACKETS.lobbyInfo, (data: PacketLobbyInfoData) => {
  io.emit("lobbyInfo", {
    players: data.m_lobbyPlayers,
    playerNumbers: data.m_numPlayers
  })
});

export default server;


src/router/index.ts
import { createRouter, createWebHistory } from 'vue-router'
// import LiveTelemetryView from '../views/LiveTelemetryView.vue'
// import ViewTelemetryView from '../views/ViewTelemetryView.vue'
import RaceTelemetryView from '../views/RaceTelemetryView.vue'
import SessionDataView from '../views/SessionDataView.vue'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    // {
    //   path: '/live',
    //   name: 'live',
    //   component: LiveTelemetryView
    // },
    // {
    //   path: '/view',
    //   name: 'view',
    //   component: ViewTelemetryView
    // },
    {
      path: '/race',
      name: 'race',
      component: RaceTelemetryView
    },
    {
      path: '/session',
      name: 'session',
      component: SessionDataView
    },
    {
      path: '/about',
      name: 'about',
      // route level code-splitting
      // this generates a separate chunk (About.[hash].js) for this route
      // which is lazy-loaded when the route is visited.
      component: () => import('../views/AboutView.vue')
    }
  ]
})

export default router


backend/app.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const f1_telemetry_client_1 = require("@racehub-io/f1-telemetry-client");
const { PACKETS } = f1_telemetry_client_1.constants;
const http_1 = __importDefault(require("http"));
const socket_io_1 = require("socket.io");
const server = http_1.default.createServer();
server.listen(3000, "0.0.0.0");
const io = new socket_io_1.Server(server, {
    cors: {
        origin: "http://localhost:5173"
    }
});
io.on('connection', (socket) => {
    console.log('a user connected');
});
const client = new f1_telemetry_client_1.F1TelemetryClient({ port: 20777, bigintEnabled: false });
client.start();
client.on(PACKETS.lapData, (data) => {
    io.emit("lapData", {
        position: data.m_lapData[data.m_header.m_playerCarIndex].m_carPosition,
        lastLapTime: data.m_lapData[data.m_header.m_playerCarIndex].m_lastLapTimeInMS,
        currentLapTime: data.m_lapData[data.m_header.m_playerCarIndex].m_currentLapTimeInMS,
        sector1Time: data.m_lapData[data.m_header.m_playerCarIndex].m_sector1TimeMinutes,
        sector2Time: data.m_lapData[data.m_header.m_playerCarIndex].m_sector2TimeMinutes,
        currentLapNum: data.m_lapData[data.m_header.m_playerCarIndex].m_currentLapNum
    });
});
client.on(PACKETS.carTelemetry, (data) => {
    io.emit("telemetryData", {
        throttle: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_throttle,
        brake: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_brake,
        gear: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_gear,
        speed: data.m_carTelemetryData[data.m_header.m_playerCarIndex].m_speed,
        frame: data.m_header.m_frameIdentifier,
    });
});
client.on(PACKETS.carStatus, (data) => {
    let tyreWearData = [];
    data.m_carStatusData.forEach((car) => {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        tyreWearData.push({
            a: (_b = (_a = car.m_tyresWear) === null || _a === void 0 ? void 0 : _a[0]) !== null && _b !== void 0 ? _b : 0,
            b: (_d = (_c = car.m_tyresWear) === null || _c === void 0 ? void 0 : _c[1]) !== null && _d !== void 0 ? _d : 0,
            c: (_f = (_e = car.m_tyresWear) === null || _e === void 0 ? void 0 : _e[2]) !== null && _f !== void 0 ? _f : 0,
            d: (_h = (_g = car.m_tyresWear) === null || _g === void 0 ? void 0 : _g[3]) !== null && _h !== void 0 ? _h : 0,
        });
    });
    io.emit('tyreWearData', {
        data: tyreWearData
    });
});
client.on(PACKETS.participants, (data) => {
    io.emit('drivers', {
        drivers: data.m_participants,
        driverCount: data.m_numActiveCars
    });
});
client.on(PACKETS.session, (data) => {
    io.emit("sessionData", {
        aiDifficulty: data.m_aiDifficulty,
        track: f1_telemetry_client_1.constants.TRACKS[data.m_trackId].name,
        weather: data.m_weather,
    });
});
client.on(PACKETS.lobbyInfo, (data) => {
    io.emit("lobbyInfo", {
        players: data.m_lobbyPlayers,
        playerNumbers: data.m_numPlayers
    });
});
exports.default = server;
